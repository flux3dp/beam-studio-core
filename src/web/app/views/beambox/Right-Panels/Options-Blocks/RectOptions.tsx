import React from 'react';

import Constant from 'app/actions/beambox/constant';
import i18n from 'helpers/i18n';
import InFillBlock from 'app/views/beambox/Right-Panels/Options-Blocks/InFillBlock';
import storage from 'implementations/storage';
import UnitInput from 'app/widgets/Unit-Input-v2';
import { getSVGAsync } from 'helpers/svg-editor-helper';

let svgCanvas;
getSVGAsync((globalSVG) => { svgCanvas = globalSVG.Canvas; });

const LANG = i18n.lang.beambox.right_panel.object_panel.option_panel;

interface Props {
  elem: Element;
  rx: number;
  updateDimensionValues: (val: { [key: string]: any }) => void;
}

function RectOptions({
  elem, rx, updateDimensionValues,
}: Props): JSX.Element {
  const handleRoundedCornerChange = (val) => {
    // eslint-disable-next-line no-param-reassign
    val *= Constant.dpmm;
    svgCanvas.changeSelectedAttribute('rx', val, [elem]);
    updateDimensionValues({ rx: val });
  };

  const renderRoundCornerBlock = () => {
    const unit = storage.get('default-units') || 'mm';
    const isInch = unit === 'inches';
    return (
      <div className="option-block" key="rounded-corner">
        <div className="label">{LANG.rounded_corner}</div>
        <UnitInput
          min={0}
          unit={isInch ? 'in' : 'mm'}
          className={{ 'option-input': true }}
          defaultValue={rx / Constant.dpmm || 0}
          getValue={(val) => handleRoundedCornerChange(val)}
        />
      </div>
    );
  };

  return (
    <div className="rect-options">
      {renderRoundCornerBlock()}
      <InFillBlock elem={elem} />
    </div>
  );
}

export default RectOptions;
