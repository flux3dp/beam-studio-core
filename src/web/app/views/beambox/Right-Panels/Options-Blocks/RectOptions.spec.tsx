/* eslint-disable import/first */
import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

jest.mock('helpers/i18n', () => ({
  lang: {
    beambox: {
      right_panel: {
        object_panel: {
          option_panel: {
            rounded_corner: 'Rounded corner',
          },
        },
      },
    },
  },
}));

const get = jest.fn();
jest.mock('implementations/storage', () => ({
  get,
}));

const getSVGAsync = jest.fn();
jest.mock('helpers/svg-editor-helper', () => ({
  getSVGAsync,
}));

const changeSelectedAttribute = jest.fn();
getSVGAsync.mockImplementation((callback) => {
  callback({
    Canvas: {
      changeSelectedAttribute,
    },
  });
});

import RectOptions from './RectOptions';

describe('should render correctly', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test('unit is inches', () => {
    get.mockReturnValue('inches');
    const updateDimensionValues = jest.fn();
    document.body.innerHTML = '<div id="flux" />';
    const wrapper = shallow(
      <RectOptions
        elem={document.getElementById('flux')}
        rx={0}
        updateDimensionValues={updateDimensionValues}
      />,
    );

    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.find('UnitInput').props().getValue(10);
    expect(changeSelectedAttribute).toHaveBeenCalledTimes(1);
    expect(changeSelectedAttribute).toHaveBeenNthCalledWith(1, 'rx', 100, [document.getElementById('flux')]);
    expect(updateDimensionValues).toHaveBeenCalledTimes(1);
    expect(updateDimensionValues).toHaveBeenNthCalledWith(1, { rx: 100 });
  });

  test('unit is not inches', () => {
    get.mockReturnValue(null);
    document.body.innerHTML = '<div id="flux" />';
    const wrapper = shallow(
      <RectOptions
        elem={document.getElementById('flux')}
        rx={10}
        updateDimensionValues={jest.fn()}
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
