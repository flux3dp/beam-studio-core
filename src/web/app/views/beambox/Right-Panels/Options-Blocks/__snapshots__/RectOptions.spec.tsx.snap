// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly unit is inches 1`] = `
<div>
  <div
    class="rect-options"
  >
    <div
      class="option-block"
    >
      <div
        class="label"
      >
        Rounded corner
      </div>
      <div
        class="option-input ui ui-control-unit-input-v2"
      >
        <input
          step="1"
          type="text"
          value="0.0000"
        />
        <span
          class="unit"
        >
          in
        </span>
      </div>
    </div>
    <div>
      DummyInFillBlock
    </div>
  </div>
</div>
`;

exports[`should render correctly unit is not inches 1`] = `
<div>
  <div
    class="rect-options"
  >
    <div
      class="option-block"
    >
      <div
        class="label"
      >
        Rounded corner
      </div>
      <div
        class="option-input ui ui-control-unit-input-v2"
      >
        <input
          step="1"
          type="text"
          value="1.00"
        />
        <span
          class="unit"
        >
          mm
        </span>
      </div>
    </div>
    <div>
      DummyInFillBlock
    </div>
  </div>
</div>
`;
